source/core/slang-uint-set.h

```diff
@@ -6,22 +6,72 @@
 #include "slang-common.h"
 #include "slang-hash.h"
 
+#if defined(_MSC_VER)
+#include <intrin.h>
+#endif
 #include <memory.h>
 
 namespace Slang
 {
 
+template<typename T>
+constexpr static Index computeElementShift()
+{
+    Index currentShift = 0;
+    Index currentShiftValue = 1;
+
+    while (currentShiftValue != sizeof(T)*8)
+    {
+        currentShift++;
+        currentShiftValue *= 2;
+    }
+
+    return currentShift;
+}
+
+static inline Index bitscanForward(const uint64_t& in)
```

Comment:
Even for composite types, if we know the function may copy the parameter internally, then we shouldn't pass by `const &`, as this will actually prevent copy ellision through the function boundary.

Reply:
I didn't know about the term, `copy elision`.
Thank for the information.
This is one of my TIL.

And it is totally right that the compiler will have a harder time optimizing the code when a given parameter is a pointer to the outside of the function that could be modified while running the function.
It is sometimes unnecessary to use a reference type to avoid a copy-construct, but I guess we can use `__restricted` for the case.

Reply:
I suppose a move is possible if a variable is never used again in the scope, there should not have been a const& here.

side note about `restrict`, I thought it was unused by clang/msvc?